### PR TITLE
virtctl, pause: Refactor commands

### PR DESCRIPTION
**Pull request by oshoval**
### What this PR does
Split commands to files.
Each file is responsible for one command only, ease readability and 
reduces collision domain.

Fixes #

### Why we need it and why it was done in this way
The following tradeoffs were made:

The following alternatives were considered:

Links to places where the discussion took place: 

### Checklist

This checklist is not enforcing, but it's a reminder of items that could be relevant to every PR.
Approvers are expected to review this list.

- [ ] Design: A [design document](https://github.com/kubevirt/community/tree/main/design-proposals) was considered and is present (link) or not required
- [ ] PR: The PR description is expressive enough and will help future contributors
- [ ] Code: [Write code that humans can understand](https://en.wikiquote.org/wiki/Martin_Fowler#code-for-humans) and [Keep it simple](https://en.wikipedia.org/wiki/KISS_principle)
- [ ] Refactor: You have [left the code cleaner than you found it (Boy Scout Rule)](https://learning.oreilly.com/library/view/97-things-every/9780596809515/ch08.html)
- [ ] Upgrade: Impact of this change on upgrade flows was considered and addressed if required
- [ ] Testing: New code requires [new unit tests](https://github.com/kubevirt/kubevirt/blob/main/docs/reviewer-guide.md#when-is-a-pr-good-enough). New features and bug fixes require at least on e2e test
- [ ] Documentation: A [user-guide update](https://github.com/kubevirt/user-guide/) was considered and is present (link) or not required. You want a user-guide update if it's a user facing feature / API change.
- [ ] Community: Announcement to [kubevirt-dev](https://groups.google.com/g/kubevirt-dev/) was considered

### Release note

```release-note
None
```

### CODE REVIEW

**Review by 0xFelix — COMMENTED** [review]
This already goes into the right direction, but ideally I'd like to see pause and unpause being split into two separate pkgs. (`pkg/virtctl/pause` and `pkg/virtctl/unpause`) Can you do that please?

**Comment by oshoval** [comment]
> This already goes into the right direction, but ideally I'd like to see pause and unpause being split into two separate pkgs. (`pkg/virtctl/pause` and `pkg/virtctl/unpause`) Can you do that please?

sounds overkill imo, unless you insist

**Comment by 0xFelix** [comment]
IMO to get a better understanding of responsibilites each command should be split into its own pkg. You already identified how the same function shares responsibilites for two commands and there are more cases of this in this file. I'd like to make commands short and concise by ensuring they live in their own (sub)packages.

**Comment by oshoval** [comment]
the whole feature is already named pause (and holds the sub package name) `pkg/virtctl/pause/pause.go`
so i will introduce `pkg/virtctl/pause/pause_cmd/pause_cmd.go` and `pkg/virtctl/pause/unpause_cmd/unpause_cmd.go`
is it ok ?

**Comment by 0xFelix** [comment]
I'd like to see it split into `pkg/virtctl/pause` and `pkg/virtctl/unpause`.

**Comment by oshoval** [comment]
> I'd like to see it split into `pkg/virtctl/pause` and `pkg/virtctl/unpause`.

but then where will be the main "shared" code that we have now on `pkg/virtctl/pause` ?
it will stay on `pkg/virtctl/pause` ?

**Comment by 0xFelix** [comment]
My point is, I don't see why there should be any shared code between these two commands.

**Comment by oshoval** [comment]
rebased

drafted, not ready yet

**Comment by oshoval** [comment]
Ready for review, thanks
(lets see all pass, locally it did)

File: pkg/virtctl/unpause/unpause_suite_test.go
Comment (0xFelix) [inline pkg/virtctl/unpause/unpause_suite_test.go:1]: Looks like an equivalent file is missing for `pause`.

File: pkg/virtctl/pause/pause.go [hunk 1]
```diff
@@ -38,23 +38,20 @@ import (
...
-	COMMAND_UNPAUSE = "unpause"
-	ARG_VM_SHORT    = "vm"
-	ARG_VM_LONG     = "virtualmachine"
-	ARG_VMI_SHORT   = "vmi"
-	ARG_VMI_LONG    = "virtualmachineinstance"
+	COMMAND_PAUSE = "pause"
+	ARG_VM_SHORT  = "vm"
+	ARG_VM_LONG   = "virtualmachine"
+	ARG_VMI_SHORT = "vmi"
+	ARG_VMI_LONG  = "virtualmachineinstance"
```
Comment (0xFelix) [inline pkg/virtctl/pause/pause.go:45]: Inline these or at least move them to the func where they are used in and don't make them exported.

File: pkg/virtctl/pause/pause.go [hunk 2]
```diff
@@ -76,33 +73,9 @@ Second argument is the name of the resource.`,
...
-	cmd.SetUsageTemplate(templates.UsageTemplate())
-	return cmd
-}
-
-func usage(cmd string) string {
-	usage := fmt.Sprintf("  # %s a virtualmachine called 'myvm':\n", strings.Title(cmd))
-	usage += fmt.Sprintf("  {{ProgramName}} %s vm myvm", cmd)
+func usage() string {
+	usage := fmt.Sprintf("  # %s a virtualmachine called 'myvm':\n", strings.Title(COMMAND_PAUSE))
+	usage += fmt.Sprintf("  {{ProgramName}} %s vm myvm", COMMAND_PAUSE)
```
Comment (0xFelix) [inline pkg/virtctl/pause/pause.go:78]: I'd drop the `fmt.Sprintf` calls and simply return a raw string block.

```
func usage() string {
    return `...`
}
```

File: pkg/virtctl/pause/pause.go [hunk 3]
```diff
@@ -124,63 +97,45 @@ func (vc *virtCommand) Run(args []string) error {
...
+		err = client.VirtualMachineInstance(namespace).Pause(context.Background(), vmiName, &kubevirtV1.PauseOptions{DryRun: dryRunOption})
+		if err != nil {
+			if errors.IsNotFound(err) {
+				runningStrategy, err := vm.RunStrategy()
+				if err != nil {
+					return fmt.Errorf("Error pausing VirtualMachineInstance %s: %v", vmiName, err)
+				}
+				if runningStrategy == kubevirtV1.RunStrategyHalted {
+					return fmt.Errorf("Error pausing VirtualMachineInstance %s. VirtualMachine %s is not set to run", vmiName, vm.Name)
+				}
```
Comment (0xFelix) [inline pkg/virtctl/pause/pause.go:121]: This validation should be covered by the API already. We shouldn't be doing any additional validations in virtctl, the API knows best.

If you drop this validation the code paths for VM/VMI could be unified and we're left with only a call to `client.VirtualMachineInstance(namespace).Pause`, which would be fine to do from `Run` and dropping this func.

File: pkg/virtctl/pause/pause.go [hunk 4]
```diff
@@ -124,63 +97,45 @@ func (vc *virtCommand) Run(args []string) error {
...
+		printLog(resourceName)
 	}
+
 	return nil
 }
 
-func printLog(resource string, command string) (int, error) {
-	return fmt.Printf("VMI %s was scheduled to %s\n", resource, command)
+func printLog(resource string) (int, error) {
+	return fmt.Printf("VMI %s was scheduled to %s\n", resource, COMMAND_PAUSE)
```
Comment (0xFelix) [inline pkg/virtctl/pause/pause.go:140]: Inline this and move it into `Run`. Also, ideally make it use `cmd.Printf`.

File: pkg/virtctl/root.go [hunk 5]
```diff
@@ -100,7 +101,7 @@ func NewVirtctlCommand() (*cobra.Command, clientcmd.ClientConfig) {
 		vm.NewExpandCommand(clientConfig),
 		memorydump.NewMemoryDumpCommand(clientConfig),
 		pause.NewPauseCommand(clientConfig),
-		pause.NewUnpauseCommand(clientConfig),
+		unpause.NewUnpauseCommand(clientConfig),
```
Comment (0xFelix) [inline pkg/virtctl/root.go:104]:
```suggestion
		unpause.NewCommand(clientConfig),
```

No need to repeat ourselves.

File: pkg/virtctl/unpause/unpause.go
Comment (0xFelix) [inline pkg/virtctl/unpause/unpause.go:1]: Comments made in `pause.go` also apply to this file.

File: pkg/virtctl/unpause/unpause_test.go [hunk 6]
```diff
@@ -0,0 +1,89 @@
...
+		kubecli.MockKubevirtClientInstance = kubecli.NewMockKubevirtClient(ctrl)
+		vmInterface = kubecli.NewMockVirtualMachineInterface(ctrl)
+		vmiInterface = kubecli.NewMockVirtualMachineInstanceInterface(ctrl)
+	})
+
+	Context("With missing input parameters", func() {
+		It("should fail an unpause", func() {
+			cmd := clientcmd.NewRepeatableVirtctlCommand(unpause.COMMAND_UNPAUSE)
+			err := cmd()
+			Expect(err).To(HaveOccurred())
```
Comment (0xFelix) [inline pkg/virtctl/unpause/unpause_test.go:42]:
```suggestion
			Expect(cmd()).To(MatchError(ContainSubstring(...)))
```

Please be a bit more specific when asserting the error.

**Review by 0xFelix — COMMENTED** [review]
Thanks, this goes into the right direction. I added some comments.

File: pkg/virtctl/pause/pause.go (hunk 3 above)
Comment (oshoval) [reply pkg/virtctl/pause/pause.go:121]: can we please limit this PR scope to the splitting ?
i must work on other stuff and cant dedicate much for it atm

File: pkg/virtctl/unpause/unpause_suite_test.go
Comment (0xFelix) [reply pkg/virtctl/unpause/unpause_suite_test.go:1]: Please address this comment, so we can merge this PR as an intermediate step in the virtctl pause/unpause cleanup.

Comment (oshoval) [reply pkg/virtctl/unpause/unpause_suite_test.go:1]: There is one `pkg/virtctl/pause/pause_suite_test.go` simply wasn't changed so it doesnt appear as we spoke

Comment (0xFelix) [reply pkg/virtctl/unpause/unpause_suite_test.go:1]: Alright, I missed that.

**Review by 0xFelix — COMMENTED** [review]
/approve

File: pkg/virtctl/pause/pause.go (hunk 4 above)
Comment (oshoval) [reply pkg/virtctl/pause/pause.go]: done the first half of the sentence

**Comment by oshoval** [comment]
marked the comments i addressed
thanks

File: pkg/virtctl/unpause/unpause_test.go (hunk 6 above)
Comment (oshoval) [reply pkg/virtctl/unpause/unpause_test.go:44]: will be done as follow-up as discussed
(same as https://github.com/kubevirt/kubevirt/pull/12665#discussion_r1749740180)

**Comment by oshoval** [comment]
@jcanocan 
can you take a look please ?

File: pkg/virtctl/unpause/unpause.go [hunk 7]
```diff
@@ -0,0 +1,115 @@
...
+			return fmt.Errorf("Error unpausing VirtualMachineInstance %s: %v", vmiName, err)
+		}
+		fmt.Printf("VMI %s was scheduled to unpause\n", vmiName)
+	case "virtualmachineinstance", "vmi":
+		err := client.VirtualMachineInstance(namespace).Unpause(context.Background(), resourceName, &kubevirtV1.UnpauseOptions{DryRun: dryRunOption})
+		if err != nil {
+			return fmt.Errorf("Error unpausing VirtualMachineInstance %s: %v", resourceName, err)
+		}
+		fmt.Printf("VMI %s was scheduled to unpause\n", resourceName)
+	}
```
Comment (jcanocan) [inline pkg/virtctl/unpause/unpause.go:112]: Reduce the calls to `VirtualMachineInstance` to just one outside the switch case statement.

**Review by jcanocan — COMMENTED** [review]
Good job!
Just a silly comment!

File: pkg/virtctl/unpause/unpause.go (hunk 7 above)
Comment (oshoval) [reply pkg/virtctl/unpause/unpause.go:112]: It doesnt have the same input params,
Moreover what if theroretically resourceType is not one of the 4 that appear in the cases,
in this event we should not do anything, and putting it on the top would do without checking those cases.
Also this PR scope was agreed to split for comments, so i prefer to not change such logic

Comment (jcanocan) [reply pkg/virtctl/unpause/unpause.go:112]: All right! Thanks for the additional context.

**Pull request merged**